roles/jupyterhub/tasks/install.yml

```diff
@@ -0,0 +1,69 @@
+- name: "Set 'nodejs_install: True' and 'nodejs_enabled: True'"
+  set_fact:
+    nodejs_install: True
+    nodejs_enabled: True
+
+- name: NODEJS - run 'nodejs' role (attempt to install & enable Node.js)
+  include_role:
+    name: nodejs
+
+- name: FAIL (STOP THE INSTALL) IF 'nodejs_installed is undefined'
+  fail:
+    msg: "JupyterHub install cannot proceed, as Node.js is not installed."
+  when: nodejs_installed is undefined
+
+
+- name: Make the directories to hold JupyterHub config
+  file:
+    state: directory
+    path: '{{ item }}'
+  with_items:
+    - '{{ jupyterhub_venv }}/etc/jupyter'
+    - '{{ jupyterhub_venv }}/etc/jupyterhub'
+    - '{{ jupyterhub_venv }}/etc/systemd'
+
+- name: Use npm to install configurable-http-proxy
+  npm:
+    name: configurable-http-proxy
+    global: yes
+    state: latest
+
+- name: 'Use pip to install into a virtual environment: {{ jupyterhub_venv }}'
+  pip:
+    name:
+      - pip
```

Comment:
Interesting that pip is installing pip (as circular as that appears!)

@georgejhunt do you happen to know how JupyterHub uses this?

Total Aside: pip3 is often installed by apt package `python3-pip` in 2021.

Reply:
That is a pretty snide comment, guess you just misunderstood that pip is installing needed dependencies into a blank isolated virtual environment where there is nothing, apart from where python3-pip lives. Can't recall off the top of my head if pip is auto-installed when the virtual environment is created or not but PyPI should install all the declared dependencies   but just being explicit is a0k in my books also. The other way would be to reference the project's requirements.txt file if one exists with a github source(calibre-web).